src/twisted/conch/client/direct.py

```diff
@@ -1,51 +1,77 @@
 # Copyright (c) Twisted Matrix Laboratories.
 # See LICENSE for details.
 
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Callable
 
 from twisted.conch import error
 from twisted.conch.ssh import transport
 from twisted.internet import defer, protocol, reactor
+from twisted.internet.address import IPv4Address, IPv6Address
+from twisted.internet.defer import Deferred, maybeDeferred
+from twisted.internet.interfaces import (
+    IAddress,
+    IConnector,
+    IListeningPort,
+    IReactorTCP,
+)
+from twisted.python.failure import Failure
+
+if TYPE_CHECKING:
+    from twisted.conch.client.options import ConchOptions
+    from twisted.conch.ssh.userauth import SSHUserAuthClient
 
 
 class SSHClientFactory(protocol.ClientFactory):
-    def __init__(self, d, options, verifyHostKey, userAuthObject):
-        self.d = d
+    def __init__(
+        self,
+        d: Deferred[None],
+        options: ConchOptions,
+        verifyHostKey: _VHK,
+        userAuthObject: SSHUserAuthClient,
+    ) -> None:
+        self.d: Deferred[None] | None = d
         self.options = options
         self.verifyHostKey = verifyHostKey
         self.userAuthObject = userAuthObject
 
-    def clientConnectionLost(self, connector, reason):
+    def clientConnectionLost(self, connector: IConnector, reason: Failure) -> None:
         if self.options["reconnect"]:
             connector.connect()
 
-    def clientConnectionFailed(self, connector, reason):
+    def clientConnectionFailed(self, connector: IConnector, reason: Failure) -> None:
         if self.d is None:
             return
         d, self.d = self.d, None
         d.errback(reason)
 
-    def buildProtocol(self, addr):
+    def buildProtocol(self, addr: IAddress) -> SSHClientTransport:
         trans = SSHClientTransport(self)
         if self.options["ciphers"]:
             trans.supportedCiphers = self.options["ciphers"]
         if self.options["macs"]:
             trans.supportedMACs = self.options["macs"]
         if self.options["compress"]:
-            trans.supportedCompressions[0:1] = ["zlib"]
+            trans.supportedCompressions[0:1] = [b"zlib"]
         if self.options["host-key-algorithms"]:
             trans.supportedPublicKeys = self.options["host-key-algorithms"]
         return trans
 
 
 class SSHClientTransport(transport.SSHClientTransport):
-    def __init__(self, factory):
+    # pre-mypy LSP violation
+    factory: SSHClientFactory  # type:ignore[assignment]
+
+    def __init__(self, factory: SSHClientFactory) -> None:
         self.factory = factory
-        self.unixServer = None
+        self.unixServer: None | IListeningPort = None
 
-    def connectionLost(self, reason):
+    def connectionLost(self, reason: Failure | None = None) -> None:
         if self.unixServer:
-            d = self.unixServer.stopListening()
-            self.unixServer = None
+            # https://github.com/twisted/twisted/issues/12418
```

Comment:
Minor comment

I think is best to also add a small text, not only the link.
It should give a preview ... and if you want to learn more, you can check the GitHub Issue

```suggestion
            # TODO: We might want to deprecated C{unixServer}
            # https://github.com/twisted/twisted/issues/12418
```

Reply:
I definitely don't want to add a "TODO" here for the reasons given above, and "deprecated" is the wrong conjugation, but I can add a brief explanation